planner/core/flat_plan.go

```diff
@@ -55,22 +55,34 @@ type FlatPhysicalPlan struct {
 type FlatPlanTree []*FlatOperator
 
 // GetSelectPlan skips Insert, Delete and Update at the beginning of the FlatPlanTree.
+// GetSelectPlan also skips foreign key check/cascade plan.
 // Note:
 //
-//	It returns a reference to the original FlatPlanTree, please avoid modifying the returned value.
+//	It returns a reference to the original FlatPlanTree and the offset, please avoid modifying the returned value.
 //	Since you get a part of the original slice, you need to adjust the FlatOperator.Depth and FlatOperator.ChildrenIdx when using them.
-func (e FlatPlanTree) GetSelectPlan() FlatPlanTree {
+func (e FlatPlanTree) GetSelectPlan() (FlatPlanTree, int) {
 	if len(e) == 0 {
-		return nil
+		return nil, 0
 	}
+	hasDML := false
 	for i, op := range e {
 		switch op.Origin.(type) {
 		case *Insert, *Delete, *Update:
+			hasDML = true
 		default:
-			return e[i:]
+			if hasDML {
```

Comment:
Is this check necessary?

Reply:
Since only DML contains foreign key check/cascade, so it is a tiny optimization to avoid always iterator all FlatPlan.